kernel/constant_typing.ml

```diff
@@ -209,10 +211,13 @@ let infer_opaque ~sec_univs env entry =
     cook_flags = Environ.typing_flags env;
   }, context
 
-let check_section_variables env declared_set typ body =
-  let ids_typ = global_vars_set env typ in
+(* Checks the section variables for the body.
+   The [declared_set] is expected to be the closure of the variables
+   appearing in the type and any additional variables.
+*)
+let check_section_variables env declared_set body =
```

Comment:
I find the API not super robust. Instead of conflating the variables from the type and the additional one, you could just have taken an optional additional set of variables that you could have added to the union below. This would have ensured that we cannot miss the invariant written in the comment.

Reply:
Do you mean something like this newest commit?

The desired behaviour when the higher layers give insufficient variables for the type is unclear, currently the kernel does the completion.